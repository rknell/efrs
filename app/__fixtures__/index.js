import users from './users.fixtures'

export default {
  users
}
