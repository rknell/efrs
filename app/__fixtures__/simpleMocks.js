
export class MockFormController {
  model = {
    disableScroll: false
  }
}
