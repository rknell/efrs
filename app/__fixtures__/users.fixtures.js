function ObjectId (data) {
  return data
}

function ISODate (data) {
  return data
}

export default {
  setupUser: {
    '_id': ObjectId('5a36fa6aac023a4b3bae2f22'),
    'createdAt': ISODate('2017-12-17T23:14:50.373Z'),
    'updatedAt': ISODate('2017-12-17T23:14:50.390Z'),
    'password': '$2a$08$gtSMlrmdhH0zOeeYJzPh.egq2I65fD9WFV5PypJfNWqikrZOzFQJq',
    'username': 'user12@example.com',
    'email': 'user12@example.com',
    'domain': 'example.com.au',
    'firstName': 'Admin',
    'lastName': 'User',
    'relationships': [],
    'pushTokens': [],
    'isSetup': true,
    '__v': 0,
    'company': ObjectId('5a36fa6aac023a4b3bae2f20')
  },
  newUser: {
    '_id': ObjectId('5a36fa6aac023a4b3bae2f22'),
    'createdAt': ISODate('2017-12-17T23:14:50.373Z'),
    'updatedAt': ISODate('2017-12-17T23:14:50.390Z'),
    'password': '$2a$08$gtSMlrmdhH0zOeeYJzPh.egq2I65fD9WFV5PypJfNWqikrZOzFQJq',
    'username': 'user12@example.com',
    'email': 'user12@example.com',
    'domain': 'example.com.au',
    'firstName': 'Admin',
    'lastName': 'User',
    'relationships': [],
    'pushTokens': [],
    'isSetup': false,
    '__v': 0,
    'company': ObjectId('5a36fa6aac023a4b3bae2f20')
  }
}
