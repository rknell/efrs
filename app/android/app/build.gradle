apply plugin: "com.android.application"


import com.android.build.OutputFile

/**
 * The react.gradle file registers a task for each build variant (e.g. bundleDebugJsAndAssets
 * and bundleReleaseJsAndAssets).
 * These basically call `react-native bundle` with the correct arguments during the Android build
 * cycle. By default, bundleDebugJsAndAssets is skipped, as in debug/dev mode we prefer to load the
 * bundle directly from the development server. Below you can see all the possible configurations
 * and their defaults. If you decide to add a configuration block, make sure to add it before the
 * `apply from: "../../node_modules/react-native/react.gradle"` line.
 *
 * project.ext.react = [
 *   // the name of the generated asset file containing your JS bundle
 *   bundleAssetName: "index.android.bundle",
 *
 *   // the entry file for bundle generation
 *   entryFile: "index.android.js",
 *
 *   // whether to bundle JS and assets in debug mode
 *   bundleInDebug: false,
 *
 *   // whether to bundle JS and assets in release mode
 *   bundleInRelease: true,
 *
 *   // whether to bundle JS and assets in another build variant (if configured).
 *   // See http://tools.android.com/tech-docs/new-build-system/user-guide#TOC-Build-Variants
 *   // The configuration property can be in the following formats
 *   //         'bundleIn${productFlavor}${buildType}'
 *   //         'bundleIn${buildType}'
 *   // bundleInFreeDebug: true,
 *   // bundleInPaidRelease: true,
 *   // bundleInBeta: true,
 *
 *   // whether to disable dev mode in custom build variants (by default only disabled in release)
 *   // for example: to disable dev mode in the staging build type (if configured)
 *   devDisabledInStaging: true,
 *   // The configuration property can be in the following formats
 *   //         'devDisabledIn${productFlavor}${buildType}'
 *   //         'devDisabledIn${buildType}'
 *
 *   // the root of your project, i.e. where "package.json" lives
 *   root: "../../",
 *
 *   // where to put the JS bundle asset in debug mode
 *   jsBundleDirDebug: "$buildDir/intermediates/assets/debug",
 *
 *   // where to put the JS bundle asset in release mode
 *   jsBundleDirRelease: "$buildDir/intermediates/assets/release",
 *
 *   // where to put drawable resources / React Native assets, e.g. the ones you use via
 *   // require('./image.png')), in debug mode
 *   resourcesDirDebug: "$buildDir/intermediates/res/merged/debug",
 *
 *   // where to put drawable resources / React Native assets, e.g. the ones you use via
 *   // require('./image.png')), in release mode
 *   resourcesDirRelease: "$buildDir/intermediates/res/merged/release",
 *
 *   // by default the gradle tasks are skipped if none of the JS files or assets change; this means
 *   // that we don't look at files in android/ or ios/ to determine whether the tasks are up to
 *   // date; if you have any other folders that you want to ignore for performance reasons (gradle
 *   // indexes the entire tree), add them here. Alternatively, if you have JS files in android/
 *   // for example, you might want to remove it from here.
 *   inputExcludes: ["android/**", "ios/**"],
 *
 *   // override which node gets called and with what additional arguments
 *   nodeExecutableAndArgs: ["node"],
 *
 *   // supply additional arguments to the packager
 *   extraPackagerArgs: []
 * ]
 */

project.ext.react = [
        entryFile            : "index.js",
        nodeExecutableAndArgs: ["/Users/ryanknell/.nvm/versions/node/v10.15.2/bin/node"],
        bundleInRelease       : false
]

apply from: "../../node_modules/react-native/react.gradle"
apply from: "../../node_modules/react-native-code-push/android/codepush.gradle"

/**
 * Set this to true to create two separate APKs instead of one:
 *   - An APK that only works on ARM devices
 *   - An APK that only works on x86 devices
 * The advantage is the size of the APK is reduced by about 4MB.
 * Upload all the APKs to the Play Store and people will download
 * the correct one based on the CPU architecture of their device.
 */
def enableSeparateBuildPerCPUArchitecture = false

/**
 * Run Proguard to shrink the Java bytecode in release builds.
 */
def enableProguardInReleaseBuilds = false

android {
    compileSdkVersion rootProject.ext.compileSdkVersion

    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }

    buildToolsVersion rootProject.ext.buildToolsVersion
    flavorDimensions "default"

    defaultConfig {
        applicationId "io.efrs"
        minSdkVersion rootProject.ext.minSdkVersion
        targetSdkVersion rootProject.ext.targetSdkVersion
        versionCode 1070600
        versionName "1.7.6"
        multiDexEnabled true
        missingDimensionStrategy 'react-native-camera', 'mlkit'

    }
    signingConfigs {
        release {
            if (project.hasProperty('EFRS_RELEASE_STORE_FILE')) {
                storeFile file(EFRS_RELEASE_STORE_FILE)
                storePassword EFRS_RELEASE_STORE_PASSWORD
                keyAlias EFRS_RELEASE_KEY_ALIAS
                keyPassword EFRS_RELEASE_KEY_PASSWORD
            }
        }
    }
    splits {
        abi {
            reset()
            enable enableSeparateBuildPerCPUArchitecture
            universalApk false  // If true, also generate a universal APK
            include "armeabi-v7a", "x86", "arm64-v8a", "x86_64"
        }
    }
    buildTypes {
        release {
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
            signingConfig signingConfigs.release
            buildConfigField "String", "CODEPUSH_KEY", '"LhfXA7OvruVi1O6ESGF2egM7teKJ70d6bab9-cb46-4a34-ae36-57599643d17f"'
        }
        debug {
            buildConfigField "String", "CODEPUSH_KEY", '""'
        }
    }
    // applicationVariants are e.g. debug, release
    applicationVariants.all { variant ->
        variant.outputs.each { output ->
            // For each separate APK per architecture, set a unique version code as described here:
            // http://tools.android.com/tech-docs/new-build-system/user-guide/apk-splits
            def versionCodes = ["armeabi-v7a":1, "x86":2, "arm64-v8a": 3, "x86_64": 4]
            def abi = output.getFilter(OutputFile.ABI)
            if (abi != null) {  // null for the universal-debug, universal-release variants
                output.versionCodeOverride =
                        versionCodes.get(abi) * 1048576 + defaultConfig.versionCode
            }
        }
    }
}

dependencies {
    implementation 'com.android.support:multidex:1.0.3'
    implementation project(':react-native-webview')

    implementation project(':react-native-device-info')
    //implementation(project(':react-native-device-info')) {
    //    exclude group: 'com.google.android.gms'
    //}

    implementation project(':bugsnag-react-native')
    implementation project(':react-native-firebase')
    //implementation "com.google.android.gms:play-services-base:16.0.1"
    implementation ("com.google.firebase:firebase-core:16.0.8"){
        exclude group: "com.google.firebase:iid"
    }

    implementation fileTree(include: ['*.jar'], dir: 'libs')
    implementation('com.facebook.react:react-native:+') {
        exclude group: "com.android.support"
    }
    // From node_modules
    implementation project(':react-native-code-push')
    implementation project(':react-native-svg')

    implementation(project(':react-native-camera')) {
        exclude group: "com.google.android.gms"
        exclude group: "com.android.support"
        exclude group: "com.google.firebase:iid"
    }

    implementation project(':react-native-camera')
    implementation project(':react-native-vector-icons')
    implementation project(':react-native-geocoder')

    //implementation("com.android.support:appcompat-v7:${rootProject.ext.supportLibVersion}")

    //implementation("com.android.support:support-v4:${rootProject.ext.supportLibVersion}")

    //implementation "com.android.support:exifinterface:${rootProject.ext.supportLibVersion}"


    implementation ("com.google.firebase:firebase-messaging:17.6.0"){
        exclude group: "com.google.firebase:iid"
    }

    implementation project(':react-native-video')
    implementation project(':react-native-background-upload')

    compile 'com.android.support:multidex:1.0.1'

}
// Run this once to be able to run the application with BUCK
// puts all compile dependencies into folder libs for BUCK to use
task copyDownloadableDepsToLibs(type: Copy) {
    from configurations.compile
    into 'libs'
}

apply plugin: 'com.google.gms.google-services'
com.google.gms.googleservices.GoogleServicesPlugin.config.disableVersionCheck = true
