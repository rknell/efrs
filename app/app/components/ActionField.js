import React, { Component } from 'react'
import { Button } from 'react-native'
import { observer } from 'mobx-react/native'
import { observe } from 'mobx'
import FieldController from '../services/FieldController'
import EditModeWrapper from '../components/EditModeWrapper'

export class ActionFieldController extends FieldController {
  settings = [{
    value: '',
    label: 'Label',
    fieldType: 'TextFieldView',
    extraProps: {}
  }, {
    value: false,
    label: 'Read Only',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }, {
    value: false,
    label: 'Hidden',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }, {
    value: false,
    label: 'Report When Hidden',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }, {
    value: '',
    label: 'Change Status To',
    fieldType: 'TextFieldView',
    extraProps: {}
  }, {
    value: '',
    label: 'Assign To Group',
    fieldType: 'GroupPickerFieldView',
    extraProps: {
      single: true
    }
  }, {
    value: '',
    label: 'Assign To User',
    fieldType: 'UserPickerFieldView',
    extraProps: {
      single: true
    }
  }, {
    value: false,
    label: 'Unassign All',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }]

  doAction = () => {
    if (this.settingsObject['Change Status To']) this.props.formController.setStatus(this.settingsObject['Change Status To'])
    if (this.settingsObject['Assign To User']) this.props.formController.assignToUser(this.settingsObject['Assign To User'])
    if (this.settingsObject['Assign To Group']) this.props.formController.assignToGroup(this.settingsObject['Assign To Group'])
    if (this.settingsObject['Unassign All']) this.props.formController.unassignAll()
  }

  resetField (label) {
    this.props.data.settings.forEach(setting => {
      if (setting.label === label) {
        this.settings.forEach(settingInfo => {
          if (settingInfo.label === setting.label) {
            setting.value = settingInfo.value
          }
        })
      }
    })
  }

  resetOnChange (changedSetting, resetSetting) {
    this.props.data.settings.forEach(setting => {
      if (setting.label === changedSetting) {
        this.disposables.push(observe(setting, 'value', (changeObj) => {
          if (changeObj.newValue) {
            if (changeObj.newValue && changeObj.newValue.forEach) {
              // Array
              if (changeObj.newValue.length) this.resetField(resetSetting)
            } else if (changeObj.newValue) {
              this.resetField(resetSetting)
            }
          }
        }))
      }
    })
  }
}

@observer
class ActionField extends Component {
  constructor (props) {
    super(props)
    this.controller = new ActionFieldController(props)
  }

  componentWillMount () {
    this.controller.initialise()
    this.controller.resetOnChange('Assign To User', 'Assign To Group')
    this.controller.resetOnChange('Assign To User', 'Unassign All')
    this.controller.resetOnChange('Assign To Group', 'Assign To User')
    this.controller.resetOnChange('Assign To Group', 'Unassign All')
    this.controller.resetOnChange('Unassign All', 'Assign To User')
    this.controller.resetOnChange('Unassign All', 'Assign To Group')
  }

  componentWillUnmount () {
    this.controller.dispose()
  }

  render () {
    return (
      <EditModeWrapper {...this.props} onPressEdit={() => {
        this.controller.viewSettings()
      }}>
        <Button title={this.controller.settingsObject['Label']}
          disabled={this.props.editMode || this.controller.settingsObject['Read Only']}
          onPress={this.controller.doAction} />
      </EditModeWrapper>
    )
  }
}

export default ActionField
