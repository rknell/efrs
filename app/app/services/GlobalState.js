import { observable, observe, spy } from 'mobx'
import { Alert, AsyncStorage, Linking } from 'react-native'
import PushNotifications from './PushNotifications'
import Translate from '../services/Translate'

class GlobalState {
  // apiPath = 'http://192.168.1.8:1676/api'
  apiPath = 'https://www.efrs.io/api'
  // apiPath = 'http://localhost:1676/api'

  @observable model = {
    userToken: null,
    user: null,
    userPermissions: {
      permissions: {},
      groups: []
    },
    canPop: false,
    forms: [],
    cachedCalls: {
      get: {},
      post: []
    },
    isBusy: 0,
    pushToken: null
  }

  current = () => {
    return {
      user: this.model.user ? this.model.user._id : 'no user',
      group: this.model.userPermissions.groups.join(',')
    }
  }

  navigation = {
    resetTo: () => {
      console.log('RESETTING NAVIGATION')
    },
    push: () => {
      console.log('PUSHING NAVIGATION')
    }
  }

  saveSetting (key, value) {
    return AsyncStorage.setItem(`@EFRS.AppSettings:${key}`, value)
  }

  getSetting (key) {
    return AsyncStorage.getItem(`@EFRS.AppSettings:${key}`)
  }

  constructor () {
    spy(event => {
      if (event.type === 'action' || event.type === 'update') {
        // console.log(event);
      }
    })

    this.getSetting('formsData').then(data => {
      if (data) this.model.forms = JSON.parse(data)
    })

    const PNObj = new PushNotifications(this._setPushToken)
    this.pushNotifications = PNObj.messaging

    observe(this.model, 'user', () => {
      try {
        if (this.model.user && this.bugsnag) {
          this.bugsnag.setUser(this.model.user._id, `${this.model.user.firstName} ${this.model.user.lastName}`, this.model.user.email)
        }
      } catch (e) {
        console.error('could not set user for bugsnag', e)
      }
    })
  }

  _setPushToken = (token) => {
    this.model.pushToken = token
  }

  openUrl (url) {
    Linking.canOpenURL(url).then(supported => {
      if (!supported) {
        Alert.alert(Translate.do('Error'), Translate.do('Can\'t open this url on this device. Try installing an app that can handle this link type and try again.'))
      } else {
        return Linking.openURL(url)
      }
    }).catch(err => console.error('An error occurred', err))
  }
}

export default new GlobalState()
