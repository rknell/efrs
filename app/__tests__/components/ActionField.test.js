import React from 'react'
import { shallow } from 'enzyme/build/index'
import ActionField, { ActionFieldController } from '../../app/components/ActionField'
import GlobalState from '../../app/services/GlobalState'
import toJSON from 'enzyme-to-json'
import { MockFormController } from '../../__fixtures__/simpleMocks'

it('should render an action field', () => {
  const wrapper = shallow(<ActionField data={{}} formController={new MockFormController()} gs={GlobalState} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('should instantiate an action field controller', () => {
  const afc = new ActionFieldController({ test: true })
  expect(afc.settings.length).toBeGreaterThan(0)
})

it('should perform an action (change status)', () => {
  const formController = new MockFormController()
  formController.setStatus = (status) => {
    expect(status).toBe('test')
  }
  const afc = new ActionFieldController({
    data: {
      settings: [{ label: 'Change Status To', value: 'test' }]
    },
    formController
  })
  afc.initialise()
  afc.doAction()
})

it('should perform an action (assign to user)', () => {
  const formController = new MockFormController()
  formController.assignToUser = (status) => {
    expect(status).toBe('test')
  }
  const afc = new ActionFieldController({
    data: {
      settings: [{ label: 'Assign To User', value: 'test' }]
    },
    formController
  })
  afc.initialise()
  afc.doAction()
})

it('should perform an action (assign to group)', () => {
  const formController = new MockFormController()
  formController.assignToGroup = (groupId) => {
    expect(groupId).toBe('test')
  }
  const afc = new ActionFieldController({
    data: {
      settings: [{ label: 'Assign To Group', value: 'test' }]
    },
    formController
  })
  afc.initialise()
  afc.doAction()
})

it('should perform an action (unassign all)', () => {
  const formController = new MockFormController()
  formController.unassignAll = (value) => {
    expect(value).toBe(undefined)
  }
  const afc = new ActionFieldController({
    data: {
      settings: [{ label: 'Unassign All', value: 'test' }]
    },
    formController
  })
  afc.initialise()
  afc.doAction()
})

it('should reset a field', () => {
  const afc = new ActionFieldController({
    data: {
      settings: [{ label: 'Unassign All', value: 'test' }]
    }
  })
  afc.initialise()
  expect(afc.settingsObject['Unassign All']).toBe('test')
  afc.resetField('Unassign All')
  expect(afc.settingsObject['Unassign All']).toBe(false)
})
