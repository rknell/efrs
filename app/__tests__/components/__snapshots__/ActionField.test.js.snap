// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an action field 1`] = `
<EditModeWrapper
  data={
    Object {
      "settings": Array [
        Object {
          "extraProps": Object {},
          "fieldType": "TextFieldView",
          "label": "Label",
          "value": "",
        },
        Object {
          "extraProps": Object {
            "options": Array [
              Object {
                "label": "Yes",
                "value": true,
              },
              Object {
                "label": "No",
                "value": false,
              },
            ],
          },
          "fieldType": "QuickPickView",
          "label": "Read Only",
          "value": false,
        },
        Object {
          "extraProps": Object {
            "options": Array [
              Object {
                "label": "Yes",
                "value": true,
              },
              Object {
                "label": "No",
                "value": false,
              },
            ],
          },
          "fieldType": "QuickPickView",
          "label": "Hidden",
          "value": false,
        },
        Object {
          "extraProps": Object {
            "options": Array [
              Object {
                "label": "Yes",
                "value": true,
              },
              Object {
                "label": "No",
                "value": false,
              },
            ],
          },
          "fieldType": "QuickPickView",
          "label": "Report When Hidden",
          "value": false,
        },
        Object {
          "extraProps": Object {},
          "fieldType": "TextFieldView",
          "label": "Change Status To",
          "value": "",
        },
        Object {
          "extraProps": Object {
            "single": true,
          },
          "fieldType": "GroupPickerFieldView",
          "label": "Assign To Group",
          "value": "",
        },
        Object {
          "extraProps": Object {
            "single": true,
          },
          "fieldType": "UserPickerFieldView",
          "label": "Assign To User",
          "value": "",
        },
        Object {
          "extraProps": Object {
            "options": Array [
              Object {
                "label": "Yes",
                "value": true,
              },
              Object {
                "label": "No",
                "value": false,
              },
            ],
          },
          "fieldType": "QuickPickView",
          "label": "Unassign All",
          "value": false,
        },
      ],
    }
  }
  formController={
    MockFormController {
      "model": Object {
        "disableScroll": true,
      },
    }
  }
  gs={
    GlobalState {
      "_setPushToken": [Function],
      "apiPath": "https://www.efrs.io/api",
      "current": [Function],
      "model": Object {
        "cachedCalls": Object {
          "get": Object {},
          "post": Array [],
        },
        "canPop": false,
        "forms": Array [],
        "isBusy": 0,
        "pushToken": "TESTTOKEN",
        "user": null,
        "userPermissions": Object {
          "groups": Array [],
          "permissions": Object {},
        },
        "userToken": null,
      },
      "navigation": Object {
        "push": [Function],
        "resetTo": [Function],
      },
      "pushNotifications": Messaging {
        "getToken": [Function],
        "onMessage": [Function],
        "requestPermissions": [Function],
      },
    }
  }
  onPressEdit={[Function]}
>
  <Button
    disabled={false}
    onPress={[Function]}
    title=""
  />
</EditModeWrapper>
`;
